core/src/main/java/cucumber/runner/OrderedTestRunEventBus.java

```diff
@@ -23,14 +33,62 @@ public Long getTime() {
     public void send(final Event event) {
         super.send(event);
         queue(event);
-    }
 
-    private void queue(final Event event) {
-        queue.add(event);
         if (event instanceof TestRunFinished) {
-            parent.sendAll(queue);
+            parent.sendAll(flatten(queue));
             queue.clear();
         }
     }
 
+    private void queue(final Event event) {
+        if (event instanceof TestCaseEvent) {
+            if (event instanceof TestCaseStarted) {
+                List<Event> testCaseEvents = new ArrayList<Event>();
+                testCaseEvents.add(event);
+                queue.add(testCaseEvents);
+            } else {
+                queue.get(queue.size() - 1).add(event);
+            }
+        } else {
+            queue.add(singletonList(event));
+        }
+    }
+
+    private static Iterable<Event> flatten(List<List<Event>> queue) {
+        Collections.sort(queue, new Comparator<List<Event>>() {
+
+            private final List<?> weights = asList(TestRunStarted.class, TestSourceRead.class, SnippetsSuggestedEvent.class, TestCaseStarted.class, TestRunFinished.class);
+
+            @Override
+            public int compare(List<Event> o1, List<Event> o2) {
+                Event event1 = o1.get(0);
+                Event event2 = o2.get(0);
+                int x = weights.indexOf(event1.getClass());
+                int y = weights.indexOf(event2.getClass());
+
+                if (x < y) return -1;
+                else if (x > y) return 1;
+
+                if (event1 instanceof TestCaseStarted && event2 instanceof TestCaseStarted) {
+                    TestCaseStarted testCaseStarted1 = (TestCaseStarted) event1;
+                    TestCaseStarted testCaseStarted2 = (TestCaseStarted) event2;
+                    return uri(testCaseStarted1).compareTo(uri(testCaseStarted2));
+                }
+
+                return 0;
+            }
+        });
+
+        ArrayList<Event> events = new ArrayList<Event>();
+        for (List<Event> eventList : queue) {
+            events.addAll(eventList);
+        }
+
+        return events;
+    }
+
+    public static String uri(TestCaseStarted testCaseStarted1) {
+        return testCaseStarted1.testCase.getUri() + ":" + testCaseStarted1.testCase.getLine();
```

Comment:
this won't work correctly  e.g. for line 1, 9, 10 

```
public static void main(String[] args) {
    final List<String> uris = Arrays.asList("feature1.path:1", "feature1.path:9", "feature1.path:10");
    System.out.println("pre-sort: " + uris);
    Collections.sort(uris);
    System.out.println("post-sort: " + uris);
}
```
pre-sort: [feature1.path:1, feature1.path:9, feature1.path:10]
post-sort: [feature1.path:1, feature1.path:10, feature1.path:9]
```

```

Reply:
Good one. Test case needs a proper comparator.

Reply:
sorted it